### PR TITLE
[REVIEW] find_and_replace function

### PR DESCRIPTION
**Pull request by williamBlazing**
Replace function moved from old libgdf repo. It was on PR# 106

### CODE REVIEW

**Comment by kkraus14** [comment]
@mt-jones @dantegd I assume hold off on merging this until after the refactor is pushed and completed?

File: libgdf/src/bench/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#=============================================================================
```
Comment (jrhemstad) [inline libgdf/src/bench/CMakeLists.txt]: I think adding Google Bench needs to be done in a separate PR. This will also need to be made to fit @mt-jones new Cmake config.

File: libgdf/src/bench/replace/replace-benchmark.cu [hunk 2]
```diff
@@ -0,0 +1,111 @@
+/*
```
Comment (jrhemstad) [inline libgdf/src/bench/replace/replace-benchmark.cu]: Separate Google Bench into another PR.

File: libgdf/src/bench/replace/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#=============================================================================
```
Comment (jrhemstad) [inline libgdf/src/bench/replace/CMakeLists.txt]: Separate Google Bench into another PR.

**Comment by harrism** [comment]
@williamBlazing the type_dispatcher PR (#379) from @jrhemstad is now merged. Could you rework this PR to use it? It should simplify your code. :)

File: cpp/src/replace/replace.cu [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+      return (cudf::detail::unwrap(other) == cudf::detail::unwrap(element));
+    }
+
+  private:
+    T element;
+  };
+
+  template <class T>
+  __global__
+  void replace_kernel(T*                          d_col_data,
```
Comment (harrism) [inline cpp/src/replace/replace.cu]: Can you add a documentation block for this kernel function?

File: cpp/src/replace/replace.cu [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      if (found_ptr != old_values_end) {
+          auto d = thrust::distance(old_values_begin, found_ptr);
+          d_col_data[i] = d_new_values[d];
+      }
+
+      i += blockDim.x * gridDim.x;
+    }
+  }
+
+  struct replace_kernel_forwarder {
```
Comment (harrism) [inline cpp/src/replace/replace.cu]: Please add block documentation for this struct / functor.

**Comment by harrism** [comment]
@jrhemstad can you review the new changes? @williamBlazing once you resolve conflicts and fix CI errors it looks good to go. Will also need to update the CHANGELOG (new check).

File: cpp/include/cudf/functions.h [hunk 6]
```diff
@@ -852,3 +852,18 @@ gdf_error gdf_quantile_aprrox(	gdf_column*  col_in,       //input column with 0
                                 double       q,            //requested quantile in [0,1]
                                 void*        t_erased_res, //type-erased result of same type as column;
                                 gdf_context* ctxt);        //context info
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Replace elements from col according to the mapping old_values to new_values
+ * 
+ * @Param[in] col gdf_column with the data to be modified
```
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]:
```suggestion
 * @Param[in,out] col gdf_column with the data to be modified
```

File: cpp/include/cudf/functions.h [hunk 7]
```diff
@@ -852,3 +852,18 @@ gdf_error gdf_quantile_aprrox(	gdf_column*  col_in,       //input column with 0
                                 double       q,            //requested quantile in [0,1]
                                 void*        t_erased_res, //type-erased result of same type as column;
                                 gdf_context* ctxt);        //context info
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Replace elements from col according to the mapping old_values to new_values
+ *
```
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]:
```suggestion
 * For each value `old_values[i]` that is present in `col`, it will be replaced with `new_values[i]`.
 * 
```

File: cpp/src/replace/replace.cu [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  struct elements_are_equal{
+    __device__
+    elements_are_equal(T el) : element{el}
+    {
+    }
+
+    __device__
+    bool operator()(T other)
+    { 
+      return (cudf::detail::unwrap(other) == cudf::detail::unwrap(element));
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: `unwrap` is no longer necessary because the `wrapper` structs now have appropriate arithmetic and boolean operators, this includes `operator==`.

File: cpp/src/replace/replace.cu [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+  };
+
+  gdf_error find_and_replace_all(gdf_column*       col,
+                                 const gdf_column* old_values,
+                                 const gdf_column* new_values)
+  {
+    GDF_REQUIRE(col != nullptr && old_values != nullptr && new_values != nullptr, GDF_DATASET_EMPTY);
+    GDF_REQUIRE(old_values->size == new_values->size, GDF_COLUMN_SIZE_MISMATCH);
+    GDF_REQUIRE(col->dtype == old_values->dtype && col->dtype == new_values->dtype, GDF_DTYPE_MISMATCH);
+    GDF_REQUIRE(col->valid == nullptr && col->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]: Minor logic change. It's fine if a column has an allocated `valid` buffer so long as the `null_count` is zero. Also should check the same thing for `old_values` and `new_values`.
```suggestion
    GDF_REQUIRE(col->valid == nullptr || col->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
    GDF_REQUIRE(old_values->valid == nullptr || old_values->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
    GDF_REQUIRE(new_values->valid == nullptr || new_values->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
```

File: cpp/src/replace/replace.cu [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+ * limitations under the License.
+ */
+
+#include <thrust/device_ptr.h>
+#include <thrust/find.h>
+#include <thrust/execution_policy.h>
+
+#include "cudf.h"
+#include "utilities/error_utils.h"
+#include "dataframe/type_dispatcher.hpp"
```
Comment (jrhemstad) [inline cpp/src/replace/replace.cu]:
```suggestion
#include "utilities/type_dispatcher.hpp"
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -3,8 +3,10 @@
 
 ## New Features
 
+- PR #350 Add gdf_find_and_replace_all method
 - PR #455 CSV Reader: Add support for user-specified decimal point and thousands separator
 - PR #439 add `DataFrame.drop` method similar to pandas
+- PR #350 Implemented Series replace function
```
Comment (harrism) [inline CHANGELOG.md]: Please combine the two changelog lines for the same PR.

File: cpp/include/cudf/functions.h [hunk 12]
```diff
@@ -912,3 +912,20 @@ gdf_error gdf_quantile_aprrox(	gdf_column*  col_in,       //input column with 0
                                 double       q,            //requested quantile in [0,1]
                                 void*        t_erased_res, //type-erased result of same type as column;
                                 gdf_context* ctxt);        //context info
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Replace elements from `col` according to the mapping `old_values` to
+ *        `new_values`, that is, for each value `old_values[i]` that is present
```
Comment (harrism) [inline cpp/include/cudf/functions.h]:
```suggestion
 *        `new_values`, that is, replace all `old_values[i]` present in `col` with `new_values[i]`.
```

File: cpp/include/cudf/functions.h [hunk 13]
```diff
@@ -912,3 +912,20 @@ gdf_error gdf_quantile_aprrox(	gdf_column*  col_in,       //input column with 0
...
+ *        in `col`, it will be replaced with `new_values[i]`.
```
Comment (harrism) [inline cpp/include/cudf/functions.h]:
```suggestion
```
(Just minor grammar fix.)

File: cpp/src/replace/replace.cu [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+#include "utilities//type_dispatcher.hpp"
+
+namespace{ //anonymous
+
+  constexpr int BLOCK_SIZE = 256;
+
+  /* --------------------------------------------------------------------------*/
+  /** 
+   * @brief Kernel that replaces elements from `d_col_data` given the following
+   *        rule: for each value `old_values[i]` in [old_values_begin`, `old_values_end`)
```
Comment (harrism) [inline cpp/src/replace/replace.cu]:
```suggestion
   *        rule: , replace all `old_values[i]` in [old_values_begin`, `old_values_end`) present in `d_col_data` with `d_new_values[i]`.
```

File: cpp/src/replace/replace.cu [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+   *        that is present in `d_col_data`, it will be replaced with `d_new_values[i]`.
```
Comment (harrism) [inline cpp/src/replace/replace.cu]:
```suggestion

```

File: cpp/src/replace/replace.cu [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+          d_col_data[i] = d_new_values[d];
+      }
+
+      i += blockDim.x * gridDim.x;
+    }
+  }
+
+  /* --------------------------------------------------------------------------*/
+  /** 
+   * @brief Functor called by the `type_dispatcher` in order to invoke and instanciate
```
Comment (harrism) [inline cpp/src/replace/replace.cu]:
```suggestion
   * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
```

File: cpp/src/replace/replace.cu [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+  /* ----------------------------------------------------------------------------*/
+  struct replace_kernel_forwarder {
+    template <typename col_type>
+    void operator()(void*       d_col_data,
+                    size_t      nrows,
+                    const void* d_old_values,
+                    const void* d_new_values,
+                    size_t      nvalues)
+    {
+      thrust::device_ptr<const col_type> old_values_begin(static_cast<const col_type*>(d_old_values));
```
Comment (harrism) [inline cpp/src/replace/replace.cu]: This needs to use RMM. Please use the new (in branch-0.5) `rmm::device_ptr<const col_type>`.

File: cpp/src/replace/replace.cu [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+  };
+
+  gdf_error find_and_replace_all(gdf_column*       col,
+                                 const gdf_column* old_values,
+                                 const gdf_column* new_values)
+  {
+    GDF_REQUIRE(col != nullptr && old_values != nullptr && new_values != nullptr, GDF_DATASET_EMPTY);
+    GDF_REQUIRE(old_values->size == new_values->size, GDF_COLUMN_SIZE_MISMATCH);
+    GDF_REQUIRE(col->dtype == old_values->dtype && col->dtype == new_values->dtype, GDF_DTYPE_MISMATCH);
+    GDF_REQUIRE(col->valid == nullptr || col->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
```
Comment (harrism) [inline cpp/src/replace/replace.cu]: If null_count is zero isn't it OK for the valid pointer to be null?

File: cpp/src/replace/replace.cu [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+
+    return GDF_SUCCESS;
+  }
+
+} //end anonymous namespace
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Replace elements from `col` according to the mapping `old_values` to
+ *        `new_values`, that is, for each value `old_values[i]` that is present
```
Comment (harrism) [inline cpp/src/replace/replace.cu]:
```suggestion
 *        `new_values`, that is, replace each value `old_values[i]` present in `col` with `new_values[i]`.
```

File: cpp/src/replace/replace.cu [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+ *        in `col`, it will be replaced with `new_values[i]`.
```
Comment (harrism) [inline cpp/src/replace/replace.cu]:
```suggestion

```

File: cpp/src/replace/replace.cu (hunk 18 above)
Comment (harrism) [reply cpp/src/replace/replace.cu]: Nevermind...

File: cpp/src/replace/replace.cu (hunk 17 above)
Comment (williamBlazing) [reply cpp/src/replace/replace.cu]: I dont see any definitions anywhere for rmm::device_ptr

Comment (jrhemstad) [reply cpp/src/replace/replace.cu]: `rmm::device_ptr` doesn't exist. I think @harrism just got it confused with `rmm::device_vector`. `thrust::device_ptr` is fine here since it's just wrapping an existing allocation.

Comment (harrism) [reply cpp/src/replace/replace.cu]: My bad.

**Comment by harrism** [comment]
I just realized that this implements the C++ side but the Python side is still to do. Do you guys have plans for this @williamBlazing ?

**Comment by harrism** [comment]
rerun tests. Grrrrr.

**Comment by kkraus14** [comment]
run tests

**Comment by harrism** [comment]
I was waiting on CI to merge.

**Comment by harrism** [comment]
@kkraus14 are we ignoring CI failures now?

**Comment by kkraus14** [comment]
@harrism as long as the required ones are passing we can ignore the HANGS for now. If a test explicitly fails with actual output that shouldn't be ignored.

**Comment by mike-wendt** [comment]
rerun tests
